google/cloud/spanner/internal/date.cc

```diff
@@ -0,0 +1,60 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/internal/date.h"
+#include <ctime>
+#include <iomanip>
+#include <sstream>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+// All the civil-time code assumes the proleptic Gregorian calendar.
+
+namespace {
+
+// RFC3339 "full-date".
+char const kDateFormat[] = "%Y-%m-%d";
+
+}  // namespace
+
+std::string DateToString(Date d) {
+  std::tm tm;
+  tm.tm_year = static_cast<int>(d.year() - 1900);
+  tm.tm_mon = d.month() - 1;
+  tm.tm_mday = d.day();
+  std::ostringstream output;
+  output << std::put_time(&tm, kDateFormat);
+  return output.str();
+}
+
+Date DateFromString(std::string const& s) {
+  std::tm tm{};
+  std::istringstream input(s);
+  input >> std::get_time(&tm, kDateFormat);
+  if (!input || input.tellg() >= 0) {
+    // Failed to match kDateFormat, or did not consumed the entire string.
+    return Date{};
```

Comment:
Given how `Value` is currently implemented (its internal `GetValue`) function, I see what you're returning a default-constructed Date here instead of an error status. However, I'm thinking that `Value::GetValue` is going to need to change to return a `StatusOr<T>`, because each Value type could fail to convert a proto to the requested C++ type. In that case case, `Value::get<T>()` already has a `StatusOr<T>` return value, so we have room to return a nice error status, but it will need to be plumbed down into `Value::GetValue`.... but that's not yet done.

Sorry for how poorly worded and confusing I made that. Here's an issue I just filed to capture this: https://github.com/googleapis/google-cloud-cpp-spanner/issues/122

Anyway, in order to support that going forward, you might want to change these functions now to return StatusOr if they could possibly fail. That way they'll be easy to switch to the new `Value::GetValue()` model.

Thoughts?

Reply:
Done.